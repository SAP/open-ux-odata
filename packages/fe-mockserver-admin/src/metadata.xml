<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <Schema Namespace="MockServerAdmin" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityType Name="Service">
                <Key>
                    <PropertyRef Name="UrlPath"/>
                </Key>
                <Property Name="UrlPath" Type="Edm.String" Nullable="false" MaxLength="255"/>
                <Property Name="InternalName" Type="Edm.String" Nullable="true" MaxLength="100"/>
                <Property Name="EncodedPath" Type="Edm.String" Nullable="true" MaxLength="255"/>
                <Property Name="ContextIsolation" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Debug" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Watch" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="HasETag" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="RegisteredAt" Type="Edm.DateTimeOffset" Nullable="false"/>
            </EntityType>
            <EntityContainer Name="AdminContainer">
                <EntitySet Name="Services" EntityType="MockServerAdmin.Service"/>
            </EntityContainer>
        </Schema>
    </edmx:DataServices>
</edmx:Edmx>