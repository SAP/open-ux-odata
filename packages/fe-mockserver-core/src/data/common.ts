import type { Action, EntitySet, EntityType, Property } from '@sap-ux/vocabularies-types';
import type { ODataMetadata } from './metadata';
import type { FileBasedMockData, KeyDefinitions } from '../mockdata/fileBasedMockData';
import type { ILogger } from '@ui5/logger';
import type ODataRequest from '../request/odataRequest';
import type { IFileLoader } from '../index';

export interface EntitySetInterface {
    checkKeyValue(mockData: object, keyValues: object, keyName: string, keyProp?: Property): boolean;
    checkFilter(mockData: object, filterExpression: any, tenantId: string, odataRequest: ODataRequest): boolean;
    checkSearch(mockData: object, searchQueries: string[], odataRequest: ODataRequest): boolean;
    executeAction(
        actionDefinition: Action,
        actionData: object | undefined,
        odataRequest: ODataRequest,
        keys: Record<string, any>
    ): Promise<any>;
    performGET(
        keyValues: KeyDefinitions,
        asArray: boolean,
        tenantId: string,
        odataRequest: ODataRequest,
        dontClone?: boolean
    ): Promise<any>;
    performPOST(
        keyValues: KeyDefinitions,
        postData: any,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<any>;
    performPATCH(
        keyValues: KeyDefinitions,
        patchData: object,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<any>;
    performDELETE(
        keyValues: KeyDefinitions,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<void>;
    getParentEntityInterface(tenantId: string): Promise<FileBasedMockData | undefined>;
    getEntityInterface(entitySetName: string, tenantId: string): Promise<FileBasedMockData | undefined>;
    getMockData(tenantId: string): FileBasedMockData;
    isV4(): boolean;
}
export interface DataAccessInterface {
    isV4(): boolean;
    getNavigationPropertyKeys(
        data: any,
        navPropDetail: any,
        currentEntityType: EntityType,
        currentEntitySet: EntitySet | undefined,
        currentKeys: Record<string, string>,
        forCreate?: boolean
    ): Record<string, string>;
    getMockEntitySet(
        entityTypeName: string,
        generateMockData?: boolean,
        containedEntityType?: EntityType,
        containedData?: any
    ): Promise<EntitySetInterface>;
    getData(odataRequest: ODataRequest): Promise<any>;
    getDraftRoot(keyValues: KeyDefinitions, _tenantId: string, entitySetDefinition: EntitySet): any;
    getMetadata(): ODataMetadata;
    debug: boolean;
    fileLoader: IFileLoader;
    log: ILogger;
}

/**
 *
 */
export class ExecutionError extends Error {
    statusCode: number;
    messageData: object;
    isSAPMessage: boolean;
    isCustomError = true;
    headers: Record<string, string>;

    constructor(
        message: string,
        statusCode: number,
        messageData: any,
        isSAPMessage: boolean,
        headers: Record<string, string> = {}
    ) {
        super(message);
        this.statusCode = statusCode;
        this.messageData = messageData;
        this.isSAPMessage = isSAPMessage;
        this.headers = headers;
    }
}

export function _getDateTimeOffset(isV4: boolean) {
    const date = new Date();
    return isV4 ? date.toISOString() : '/Date(' + date.getTime() + '+0000)/';
}

const CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

/**
 * Generate an ID of a given length.
 *
 * @param length
 * @returns the generated ID
 */
export function generateId(length: number) {
    let result = '';
    while (length--) {
        result += CHARS.charAt(Math.floor(Math.random() * CHARS.length));
    }
    return result;
}

export function uuidv4() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        const r = (Math.random() * 16) | 0,
            v = c == 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
