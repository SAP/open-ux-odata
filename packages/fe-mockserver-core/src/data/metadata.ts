import { parse } from '@sap-ux/edmx-parser';
import { convert } from '@sap-ux/annotation-converter';
import type {
    Action,
    ActionImport,
    ConvertedMetadata,
    EntitySet,
    EntityType,
    NavigationProperty,
    RawMetadata,
    Singleton
} from '@sap-ux/vocabularies-types';

type NameAndNav = {
    name: string;
    entitySet: EntitySet | Singleton;
    navigation: NavigationProperty;
    partner?: NavigationProperty;
};

/**
 *
 */
export class ODataMetadata {
    public static async parse(edmx: string, metadataUrl: string, ETag?: string): Promise<ODataMetadata> {
        const metadata: RawMetadata = parse(edmx);
        const converterMetadata: ConvertedMetadata = convert(metadata);
        return new ODataMetadata(converterMetadata, metadata, metadataUrl, edmx, ETag);
    }

    public getEdmx() {
        return this.edmx;
    }

    public getMetadataUrl(): string {
        return this.metadataUrl;
    }

    public readonly typeToEntityMapping: object = {};
    public readonly ETag: string | undefined;
    public readonly metadataUrl: string;
    protected readonly metadata: ConvertedMetadata;
    protected readonly parserMetadata: RawMetadata;

    private constructor(
        metadata: ConvertedMetadata,
        parserMetadata: RawMetadata,
        metadataUrl: string,
        private edmx: string,
        ETag?: string
    ) {
        this.metadata = metadata;
        this.parserMetadata = parserMetadata;
        this.metadataUrl = metadataUrl;
        this.ETag = ETag;
    }

    public getVersion(): string {
        return this.metadata.version === '1.0' ? '2.0' : '4.0';
    }

    public getEntitySet(entitySetName: string): EntitySet | undefined {
        return this.metadata.entitySets.find((entitySet) => entitySet.name === entitySetName);
    }

    public getSingleton(singletonName: string): EntitySet {
        return this.metadata.singletons.find((singleton) => singleton.name === singletonName) as any;
    }

    public getEntitySets(): EntitySet[] {
        return this.metadata.entitySets;
    }

    public getSingletons(): Singleton[] {
        return this.metadata.singletons || [];
    }

    public getEntityType(entityTypeName: string): EntityType | undefined {
        return this.metadata.entityTypes.find((entityType) => entityType.name === entityTypeName);
    }

    public getEntitySetByType(entityTypeName: string): EntitySet | undefined {
        return this.metadata.entitySets.find((entitySet) => entitySet.entityTypeName === entityTypeName);
    }

    public getActionByFQN(actionFQN: string): Action | undefined {
        return this.metadata.actions.find((action) => action.fullyQualifiedName === actionFQN);
    }

    public getActionImportByFQN(actionImportFQN: string): ActionImport | undefined {
        return this.metadata.actionImports.find((actionImport) => actionImport.fullyQualifiedName === actionImportFQN);
    }

    public getEntityContainerPath(): string {
        return this.parserMetadata.schema.entityContainer.fullyQualifiedName;
    }

    public isDraftEntity(entitySet: EntitySet): boolean {
        return (
            entitySet?.annotations?.Common?.DraftRoot !== undefined ||
            entitySet?.annotations?.Common?.DraftNode !== undefined
        );
    }

    public isStickyEntity(entitySet: EntitySet): boolean {
        return entitySet?.annotations?.Session?.StickySessionSupported !== undefined;
    }

    public isDraftRoot(entitySet: EntitySet): boolean {
        return entitySet?.annotations?.Common?.DraftRoot !== undefined;
    }

    public isDraftNode(entitySet: EntitySet | Singleton): boolean {
        return entitySet._type === 'EntitySet' && entitySet.annotations.Common?.DraftNode !== undefined;
    }

    public resolvePath(path: string): any {
        return this.metadata.resolvePath(path);
    }

    public findInDescendant(
        entitySet: EntitySet | Singleton,
        targetEntitySet: EntitySet,
        path: NameAndNav[],
        entitySetFilter: (entitySet: EntitySet | Singleton) => boolean = () => true
    ): boolean {
        if (entitySet === targetEntitySet) {
            // nothing to do - we are there already
            return true;
        }

        for (const [navPropBindingName, target] of Object.entries(entitySet.navigationPropertyBinding)) {
            if (!entitySetFilter(target)) {
                continue;
            }

            const navProp = entitySet.entityType.navigationProperties.find(
                (navigationProperty) => navigationProperty.name === navPropBindingName
            );

            if (path.some((segment) => segment.entitySet === target)) {
                // would add a cycle
                return false;
            } else if (navProp) {
                const partnerNavProp = navProp.targetType.navigationProperties.find(
                    (navigationProperty) => navigationProperty.name === navProp.partner
                );
                path.push({
                    name: navPropBindingName,
                    entitySet: target,
                    navigation: navProp,
                    partner: partnerNavProp
                });
            }

            if (target === targetEntitySet) {
                return true;
            } else {
                return this.findInDescendant(target, targetEntitySet, path, entitySetFilter);
            }
        }

        return false;
    }

    public resolveDraftRoot(entitySet: EntitySet) {
        if (this.isDraftEntity(entitySet)) {
            // entity set must be a draft node, otherwise there is no point in looking for its root
            const draftRoots = this.metadata.entitySets.filter(this.isDraftRoot);

            for (const draftRoot of draftRoots) {
                const path: NameAndNav[] = [];
                const found = this.findInDescendant(draftRoot, entitySet, path, this.isDraftNode);
                if (found) {
                    return {
                        found: true,
                        entitySet: draftRoot,
                        path: path
                    };
                }
            }
        }

        return { found: false, entitySet: undefined, path: [] };
    }

    public resolveAncestors(entitySet: EntitySet) {
        let found = false;
        let foundPath!: NameAndNav[];
        let rootEntitySet!: EntitySet;
        this.metadata.entitySets.forEach((et) => {
            if (!found) {
                const resolvePath: any[] = [];
                found = this.findInDescendant(et, entitySet, resolvePath);
                if (found) {
                    foundPath = resolvePath.concat();
                    rootEntitySet = et;
                }
            }
        });
        return {
            found,
            entitySet: rootEntitySet,
            path: foundPath
        };
    }

    public getParentEntitySetName(sourceEntitySet: EntitySet) {
        const { found, path, entitySet } = this.resolveAncestors(sourceEntitySet);
        if (found) {
            if (path.length === 1 && entitySet) {
                return entitySet.name;
            } else {
                return path[path.length - 1].entitySet.name;
            }
        }
    }

    public getETag() {
        return this.ETag;
    }

    public getKeys(dataLine: any, entityType: EntityType): Record<string, string | number | boolean> {
        const keys = entityType.keys;
        const keyValues: any = {};
        keys.forEach((keyProp) => {
            keyValues[keyProp.name] = dataLine[keyProp.name];
        });

        return keyValues;
    }
}
