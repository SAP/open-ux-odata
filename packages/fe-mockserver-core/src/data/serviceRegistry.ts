import type { DataAccessInterface } from './common';

/**
 * Registry for managing cross-service communication in the mockserver.
 * Allows services to access entity interfaces from other services in a controlled manner.
 */
export class ServiceRegistry {
    private readonly services: Map<string, DataAccessInterface> = new Map();

    /**
     * Register a service with its DataAccess instance
     * @param serviceName - The name/path of the service
     * @param dataAccess - The DataAccess instance for this service
     */
    public registerService(serviceName: string, dataAccess: DataAccessInterface): void {
        this.services.set(serviceName, dataAccess);
    }

    /**
     * Get a DataAccess instance for a specific service
     * @param serviceName - The name/path of the service
     * @returns The DataAccess instance or undefined if not found
     */
    public getService(serviceName: string): DataAccessInterface | undefined {
        return this.services.get(serviceName);
    }

    /**
     * Get all registered service names
     * @returns Array of service names
     */
    public getServiceNames(): string[] {
        return Array.from(this.services.keys());
    }
}
