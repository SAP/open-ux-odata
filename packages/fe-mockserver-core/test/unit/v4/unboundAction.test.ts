import { join } from 'path';
import FileSystemLoader from '../../../src/plugins/fileSystemLoader';
import CDSMetadataProvider from '@sap-ux/fe-mockserver-plugin-cds';
import { ODataMetadata } from '../../../src/data/metadata';
import { DataAccess } from '../../../src/data/dataAccess';
import type { ServiceConfig } from '../../../src';
import ODataRequest from '../../../src/request/odataRequest';

describe('Unbound Action', () => {
    let metadata, dataAccess: DataAccess;
    const baseUrl = '/test';
    const fileLoader = new FileSystemLoader();
    const metadataProvider = new CDSMetadataProvider(fileLoader);
    beforeAll(async () => {
        const baseDir = join(__dirname, 'services', 'unboundAction');
        const edmx = await metadataProvider.loadMetadata(join(baseDir, 'service.cds'));

        metadata = await ODataMetadata.parse(edmx, baseUrl + '/$metadata');
        dataAccess = new DataAccess({ mockdataPath: baseDir } as ServiceConfig, metadata, fileLoader);
    });
    test('can be called', async () => {
        const requestUrl = '/unboundAction';

        const request = new ODataRequest({ method: 'POST', url: requestUrl }, dataAccess);
        expect(request.queryPath).toMatchSnapshot();
        const actionData = await dataAccess.performAction(request);
        expect(actionData).toMatchSnapshot();
    });

    test('can call the base interface', async () => {
        const requestUrl = '/unboundActionThatFetchData';

        const request = new ODataRequest({ method: 'POST', url: requestUrl }, dataAccess);
        expect(request.queryPath).toMatchSnapshot();
        const actionData = await dataAccess.performAction(request);
        expect(actionData).toMatchSnapshot();
    });

    test('can call the base interface for an incorrect entity   ', async () => {
        const requestUrl = '/unboundActionThatFetchDataOnUnknownEntity';

        const request = new ODataRequest({ method: 'POST', url: requestUrl }, dataAccess);
        expect(request.queryPath).toMatchSnapshot();
        expect(async () => await dataAccess.performAction(request)).rejects.toThrow();
    });

    test('does not catch all', async () => {
        const requestUrl = '/unboundActionThatIDontknow';

        const request = new ODataRequest({ method: 'POST', url: requestUrl }, dataAccess);
        const actionData = await dataAccess.performAction(request);
        expect(actionData).toBeNull();
    });
});
