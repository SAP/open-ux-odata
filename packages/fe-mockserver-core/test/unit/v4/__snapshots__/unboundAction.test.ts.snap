// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unbound Action can be called 1`] = `
[
  {
    "keys": {},
    "path": "unboundAction",
  },
]
`;

exports[`Unbound Action can be called 2`] = `"This is my action response"`;

exports[`Unbound Action can call the base interface 1`] = `
[
  {
    "keys": {},
    "path": "unboundActionThatFetchData",
  },
]
`;

exports[`Unbound Action can call the base interface 2`] = `
[
  {
    "BooleanProperty": false,
    "Currency": "",
    "DescriptionProperty": "",
    "HasActiveEntity": true,
    "HasDraftEntity": false,
    "ID": 2,
    "IsActiveEntity": true,
    "NameProperty": "",
    "NumberProperty": 0,
    "Progress": 0,
    "TitleProperty": "This is my Title",
  },
]
`;

exports[`Unbound Action can call the base interface for an incorrect entity    1`] = `
[
  {
    "keys": {},
    "path": "unboundActionThatFetchDataOnUnknownEntity",
  },
]
`;
